### DISCUSSION TITLE
I propose to add an alternative to JS-in-CSS when building plugins.

### DISCUSSION THREAD

**Discussion by NnicanBuak** [tailwindlabs/tailwindcss#11012] [Ideas]
# What if there was an alternative to using the not-so-friendly JS-in-CSS syntax versus Utility Tailwindcss classes? It would make it very easy to create custom components and base styles inside Tailwindcss.
> Let`s make Tailwind more Utility.

What it looks like now:
```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  plugins: [
    plugin(function({ addComponents }) {
      addComponents({
        '.button': {
          padding: theme('padding.1'),
          borderRadius: theme('borderRadius.xl'),
          fontWeight: '600',
        },
        '.button-primary': {
          backgroundColor: theme(colors.primary),
          color: theme(colors.white),
          '&:hover': {
            backgroundColor: theme(colors.secondary),
          },
          dataPonterType: 'highlight'
        },
      })
    })
  ]
}
```
---
What it might look like with the new `tailwind()` plugin feature (this function does not exclude the existence of `theme()`):
```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  plugins: [
    plugin(function({ addComponents, tailwind }) {
      addComponents({
        '.button': tailwind('p-1 rounded-xl font-[600]'),
        '.button-primary': {
          tailwind('bg-primary text-white hover:bg-secondary'),
          dataPonterType: 'highlight'
          },
        },
      })
    })
  ]
}
```

## This feature is especially useful when you want to write not just styles, but styles referencing the theme(), as in the example.

**Comment by wongjn** [comment]
You can actually do this already with `@apply`, like:

```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  plugins: [
    plugin(function({ addComponents }) {
      addComponents({
        '.button': {
           '@apply p-1 rounded-xl font-[600]': {},
        },
        '.button-primary': {
          '@apply bg-primary text-white hover:bg-secondary': {},
          dataPonterType: 'highlight'
        },
      })
    })
  ]
}
```

**Reply by NnicanBuak** [reply]
Awesome, then maybe we should update the documentation to tell about this Usage Case? I missed the functions and directives and had no idea what they do. And yet the discussion doesn't end there, because it still looks like a crutch (because of the empty curly brackets), when `theme()` exists, why can't `tailwind()` exist.

**Reply by NnicanBuak** [reply]
But of course I use this crutch because I like this style better than writing CSS-in-JS tailwind styles.

**Reply by NnicanBuak** [reply]
I included this trick, but it seems to me that it doesn't work fully as a theme() because it only uses the standard class properties, not the ones changed in config.

**Reply by adamwathan** [reply]
I personally wouldn't use `@apply` in a plugin because there's no guarantee that it will actually work in someone's project, since they can override config values that would cause the classes being applied to not work.

For example if someone changes their spacing scale, `@apply p-1` might fail if there's no `1` value in the scale.

If this is just for some custom classes that are part of a specific project and not meant to be in a reusable plugin, I would just do this in CSS:

```css
@layer components {
  .button {
    @apply p-1 rounded-xl font-[600];
  }

  .button-primary {
    @apply bg-primary text-white hover:bg-secondary;
    data-pointer-type: highlight;
  }
}
```

For future versions of Tailwind we're working on some features to encourage doing more of this sort of thing in CSS vs. your config file — to me that's the more sensible place to write custom style stuff.

**Reply by NnicanBuak** [reply]
I agree, it sounds much more convenient, especially in this case, the assembly will not have to reboot to update cofigure.

**Reply by NnicanBuak** [reply]
> If this is just for some custom classes that are part of a specific project and not meant to be in a reusable plugin, I would just do this in CSS:
> 
> ```css
> @layer components {
>   .button {
>     @apply p-1 rounded-xl font-[600];
>   }
> 
>   .button-primary {
>     @apply bg-primary text-white hover:bg-secondary;
>     data-pointer-type: highlight;
>   }
> }
> ```
I tried adding the component the way suggested but it didn't work, it doesn't seem to work the same way as the original example.
![image](https://user-images.githubusercontent.com/89499782/234347243-3dde3109-a9fd-4e8a-9638-7db99fba6dc7.png)

**Reply by therealparmesh** [reply]
@adamwathan

Using `@apply` in plugins currently provides the best overall developer experience.

You can create custom components and those who implement your plugins will see those component names autocompleted via Intellisense. Also, with a tiny configuration change to the Intellisense plugin, you can get Tailwind CSS classes autocompleted in your plugin code, making development of plugins as easy as development in general.

Granted, it might not be ideal for sharing plugins to the public, but it's great for sharing plugins within a company and such, where theme values and structures would be shared alongside plugins.

**Reply by therealparmesh** [reply]
I don't think of plugins for Tailwind as too different from Babel/SWC or ESLint plugins.

Plugins can be for sharing with the general public or scoped for sharing amongst codebases with high affinity, like those within a company.

**Reply by therealparmesh** [reply]
Side note - I think you'd run into the exact same issue with CSS today if you tried to share custom components.

If you `@import` a shared CSS file that does `@apply` with configuration values that have been overridden, you'd be in the same predicament, just with different syntax.

**Reply by vigetlabs-anthony** [reply]
1. Can someone point out where in the documentation it says we can use @apply like that?
2. Can someone explain what the empty curly braces after @apply is supposed to do?

**Reply by wongjn** [reply]
1. The documentation does not say anywhere that we can use `@apply` like that.
2. The empty braces after `@apply` are to give a value to the object key. A value must be given for an object key, otherwise it would be a JavaScript syntax error.

**Reply by vigetlabs-anthony** [reply]
Thanks for confirming, I dug through the docs and looks like you're right. 

I know that the docs doesn't say that we could use @apply like that, but would you happen to know why or how @apply works like that? I did look over the CSS-in-JS section and have a general idea, but want to understand the mechanics.

**Reply by wongjn** [reply]
The CSS is generated before `@apply` rules are "expanded" by Tailwind.

**Reply by adamwathan** [reply]
> The CSS is generated before `@apply` rules are "expanded" by Tailwind.

Yep exactly, it works by coincidence not by design 🤪

**Reply by elbotho** [reply]
> For future versions of Tailwind we're working on some features to encourage doing more of this sort of thing in CSS vs. your config file — to me that's the more sensible place to write custom style stuff.

Just a quick question from the future 🔮 @adamwathan:
Is there anything planed that we can write and export CSS (instead of CSS-in-JS) from plugins? 

Right now the DX is pretty poor here (no autocomplete etc. unless I'm doing something wrong).

**Comment by NnicanBuak** [comment]
> (this function does not exclude the existence of theme())

Although, the `tailwind()` function could be an advanced version of `theme()`, but removing `theme()` is impossible.
